android/CouchbaseLite/src/androidTest/java/com/couchbase/lite/QueryTest.java

```diff
@@ -0,0 +1,379 @@
+package com.couchbase.lite;
+
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+public class QueryTest extends BaseTest {
+    private interface QueryResult {
+        void check(long n, QueryRow row) throws Exception;
+    }
+
+    private int verifyQuery(Query query, QueryResult result) throws Exception {
+        int n = 0;
+        QueryRow row;
+        ResultSet rs = query.run();
+        while ((row = rs.next()) != null) {
+            n += 1;
+            result.check(n, row);
+        }
+        return n;
+    }
+
+    private List<Map<String, Object>> loadNumbers(final int num) throws Exception {
+        final List<Map<String, Object>> numbers = new ArrayList<Map<String, Object>>();
+        db.inBatch(new Runnable() {
+            @Override
+            public void run() {
+                for (int i = 1; i <= num; i++) {
+                    Document doc = db.getDocument("doc"+i);
+                    doc.set("number1", i);
+                    doc.set("number2", num - i);
+                    doc.save();
+                    numbers.add(doc.getProperties());
+                }
+            }
+        });
+        return numbers;
+    }
+
+    private void runTestWithNumbers(List<Map<String, Object>> numbers, Object[][] cases)
+            throws Exception {
+        for (Object[] c : cases) {
+            Expression w = (Expression) c[0];
+            String[] docIDs = (String []) c[1];
+            final List<String> docIDList = new ArrayList<String>(Arrays.asList(docIDs));
+            Query q = Query.select().from(DataSource.database(db)).where(w);
+            int rows = verifyQuery(q, new QueryResult() {
+                @Override
+                public void check(long n, QueryRow row) throws Exception {
+                    String docID = row.getDocumentID();
+                    if (docIDList.contains(docID))
+                        docIDList.remove(docID);
+                }
+            });
+            assertEquals(0, docIDList.size());
+            assertEquals(docIDs.length, rows);
+        }
+    }
+
+    private String[] $docids(int...numbers) {
```

Comment:
It's just a language kimick for generating somethings. I borrowed the $ from objc.